src/audio/selector/selector_generic.c

```diff
@@ -30,16 +33,33 @@ static void sel_s16le_1ch(struct comp_dev *dev, struct audio_stream *sink,
 			  const struct audio_stream *source, uint32_t frames)
 {
 	struct comp_data *cd = comp_get_drvdata(dev);
-	int16_t *src;
-	int16_t *dest;
-	uint32_t i;
-	uint32_t j = 0;
-	uint32_t nch = cd->config.in_channels_count;
-
-	for (i = cd->config.sel_channel; i < frames * nch; i += nch) {
-		src = audio_stream_read_frag_s16(source, i);
-		dest = audio_stream_write_frag_s16(sink, j++);
-		*dest = *src;
+	int16_t *src = source->r_ptr;
+	int16_t *dest = sink->w_ptr;
+	int16_t *src_ch;
+	int nmax;
+	int i;
+	int n;
+	int processed = 0;
+	const int source_frame_bytes = audio_stream_frame_bytes(source);
+	const unsigned int nch = source->channels;
+	const unsigned int sel_channel = cd->config.sel_channel; /* 0 to nch - 1 */
+
+	assert(sel_channel < nch);
```

Comment:
lets not assert here, we should return an error with a trace.

Reply:
I'd prefer to have config check before copy() streaming starts in params or prepare. There's some checks against max channel count but I could not find this. I'll see if I can add it there and remove this assert.